File: framework/languages/java/samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/play/infrastructure/zlink/spots/entryspot/PlayEntrySpot.java
package systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.entryspot;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.spots.ZLinkEntrySpot;
import systems.zlink.framework.spots.ZLinkEntrySpotContext;
import systems.zlink.framework.spots.ZLinkActorCreateResponse;
import systems.zlink.samples.tictactoe.server.configuration.PlaySettings;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.actors.PlayActor;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.entryspot.handlers.PlayActorJoinGameHandler;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.entryspot.handlers.PlayActorObserveMilestoneHandler;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.entryspot.handlers.PlayerWinMilestoneMsgHandler;
import systems.zlink.samples.tictactoe.shared.contracts.ObserveMilestoneRes;
import systems.zlink.samples.tictactoe.shared.contracts.PlayerInfo;
import systems.zlink.samples.tictactoe.shared.contracts.PlayerWinMilestoneMsg;
import systems.zlink.samples.tictactoe.shared.contracts.WinMilestoneNotify;

// --8<-- [start:doc-entry-spot]
public final class PlayEntrySpot implements ZLinkEntrySpot<PlayActor> {
    private static final Logger LOGGER = LoggerFactory.getLogger(PlayEntrySpot.class);
    private final ZLinkEntrySpotContext context;
    private final PlaySettings settings;
    private final java.util.List<PlayActor> milestoneObservers = new java.util.ArrayList<>();

    public PlayEntrySpot(
        ZLinkEntrySpotContext context,
        PlaySettings settings) {
        this.context = context;
        this.settings = settings;
    }

    @Override
    public ZLinkEntrySpotContext context() {
        return context;
    }

    @Override
    public java.util.concurrent.CompletionStage<ZLinkActorCreateResponse> onCreateActor(
        PlayActor actor,
        ZLinkMessage createRequest) {
        if (createRequest.isEmpty()) {
            return java.util.concurrent.CompletableFuture.completedFuture(
                ZLinkActorCreateResponse.accept());
        }
        actor.applyPlayer(createRequest.decode(PlayerInfo.class));
        return java.util.concurrent.CompletableFuture.completedFuture(
            ZLinkActorCreateResponse.accept());
    }

    @Override
    public java.util.concurrent.CompletionStage<Void> onJoinedActor(PlayActor actor) {
        if (actor.destroyAfterEntrySpotJoin()) {
            return context.destroyActor(actor)
                .thenRun(() -> LOGGER.info(
                    "tictactoe actor destroy completed actor={}", actor.actorId()));
        }
        return java.util.concurrent.CompletableFuture.completedFuture(null);
    }

    @Override
    public java.util.concurrent.CompletionStage<Void> onLeaveActor(PlayActor actor) {
        milestoneObservers.removeIf(existing -> existing.actorId().equals(actor.actorId()));
        return java.util.concurrent.CompletableFuture.completedFuture(null);
    }

    @Override
    public java.util.concurrent.CompletionStage<Void> onDisconnectActor(PlayActor actor) {
        actor.markDisconnected();
        milestoneObservers.removeIf(existing -> existing.actorId().equals(actor.actorId()));
        return java.util.concurrent.CompletableFuture.completedFuture(null);
    }

    public ObserveMilestoneRes observeMilestone(PlayActor actor) {
        rememberObserver(actor);
        return new ObserveMilestoneRes(true);
    }

    public void notifyMilestone(PlayerWinMilestoneMsg event) {
        WinMilestoneNotify payload = new WinMilestoneNotify(
            event.roomId(),
            event.actorId(),
            event.displayName(),
            event.wins());
        for (PlayActor observer : java.util.List.copyOf(milestoneObservers)) {
            observer.context().boundSession()
                .send(payload)
                .submit();
        }
    }

    private void rememberObserver(PlayActor actor) {
        milestoneObservers.removeIf(existing -> existing.actorId().equals(actor.actorId()));
        milestoneObservers.add(actor);
    }
}
// --8<-- [end:doc-entry-spot]

File: framework/languages/java/samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/play/infrastructure/zlink/spots/entryspot/PlayEntrySpot.kt
package systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.spots.entryspot

import kotlinx.coroutines.future.await
import org.slf4j.LoggerFactory
import systems.zlink.framework.kotlin.addHandler
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpot
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.spots.ZLinkEntrySpotContext
import systems.zlink.framework.spots.ZLinkActorCreateResponse
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleNames
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleSettings
import systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.actors.PlayActor
import systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.spots.entryspot.handlers.PlayActorJoinGameHandler
import systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.spots.entryspot.handlers.PlayActorObserveMilestoneHandler
import systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.spots.entryspot.handlers.PlayerWinMilestoneMsgHandler
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.ObserveMilestoneRes
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlayerInfo
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlayerWinMilestoneMsg
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.WinMilestoneNotify

// --8<-- [start:doc-entry-spot]
class PlayEntrySpot(
    override val context: ZLinkEntrySpotContext,
    private val settings: SampleSettings,
) : ZLinkSuspendingEntrySpot<PlayActor>() {
    private val milestoneObservers = mutableListOf<PlayActor>()

    override suspend fun onCreateActorSuspending(
        actor: PlayActor,
        createRequest: ZLinkMessage,
    ): ZLinkActorCreateResponse {
        if (createRequest.isEmpty) {
            return ZLinkActorCreateResponse.accept()
        }
        actor.applyPlayer(createRequest.decode(PlayerInfo::class.java))
        return ZLinkActorCreateResponse.accept()
    }

    override suspend fun onJoinedActorSuspending(actor: PlayActor) {
        if (actor.destroyAfterEntrySpotJoin) {
            context.destroyActor(actor).await()
            logger.info("tictactoe actor destroy completed actor={}", actor.actorId)
        }
    }

    override suspend fun onLeaveActorSuspending(actor: PlayActor) {
        milestoneObservers.removeIf { it.actorId == actor.actorId }
    }

    override suspend fun onDisconnectActorSuspending(actor: PlayActor) {
        actor.markDisconnected()
        milestoneObservers.removeIf { it.actorId == actor.actorId }
    }

    fun observeMilestone(actor: PlayActor): ObserveMilestoneRes {
        rememberObserver(actor)
        return ObserveMilestoneRes(true)
    }

    fun notifyMilestone(event: PlayerWinMilestoneMsg) {
        val payload = WinMilestoneNotify(
            roomId = event.roomId,
            actorId = event.actorId,
            displayName = event.displayName,
            wins = event.wins,
        )
        milestoneObservers.toList().forEach { observer ->
            observer.context().boundSession()
                .send(payload)
                .submit()
        }
    }

    private fun rememberObserver(actor: PlayActor) {
        milestoneObservers.removeIf { it.actorId == actor.actorId }
        milestoneObservers += actor
    }

    private companion object {
        val logger = LoggerFactory.getLogger(PlayEntrySpot::class.java)
    }
}
// --8<-- [end:doc-entry-spot]
